File: android/src/main/kotlin/ai/annadata/mqttquic/mqtt/MQTT5Protocol.kt
package ai.annadata.mqttquic.mqtt

/**
 * MQTT 5.0 Protocol implementation. Matches MQTTD mqttd/protocol_v5.py.
 */
object MQTT5Protocol {
    
    fun buildConnectV5(
        clientId: String,
        username: String? = null,
        password: String? = null,
        keepalive: Int = 60,
        cleanStart: Boolean = true,
        sessionExpiryInterval: Int? = null,
        receiveMaximum: Int? = null,
        maximumPacketSize: Int? = null,
        topicAliasMaximum: Int? = null,
        requestResponseInformation: Int? = null,
        requestProblemInformation: Int? = null,
        authenticationMethod: String? = null,
        authenticationData: ByteArray? = null,
        properties: Map<Int, Any>? = null
    ): ByteArray {
        val variableHeader = mutableListOf<Byte>()
        variableHeader.addAll(MQTTProtocol.encodeString("MQTT").toList())
        variableHeader.add(MQTTProtocolLevel.V5)
        
        var flags = 0
        if (cleanStart) flags = flags or 0x02
        if (username != null) flags = flags or MQTTConnectFlags.USERNAME
        if (password != null) flags = flags or MQTTConnectFlags.PASSWORD
        variableHeader.add(flags.toByte())
        variableHeader.add((keepalive shr 8).toByte())
        variableHeader.add((keepalive and 0xFF).toByte())
        
        val connectProps = mutableMapOf<Int, Any>()
        sessionExpiryInterval?.let { connectProps[MQTT5PropertyType.SESSION_EXPIRY_INTERVAL.toInt()] = it }
        receiveMaximum?.let { connectProps[MQTT5PropertyType.RECEIVE_MAXIMUM.toInt()] = it }
        maximumPacketSize?.let { connectProps[MQTT5PropertyType.MAXIMUM_PACKET_SIZE.toInt()] = it }
        topicAliasMaximum?.let { connectProps[MQTT5PropertyType.TOPIC_ALIAS_MAXIMUM.toInt()] = it }
        requestResponseInformation?.let { connectProps[MQTT5PropertyType.REQUEST_RESPONSE_INFORMATION.toInt()] = it }
        requestProblemInformation?.let { connectProps[MQTT5PropertyType.REQUEST_PROBLEM_INFORMATION.toInt()] = it }
        authenticationMethod?.let { connectProps[MQTT5PropertyType.AUTHENTICATION_METHOD.toInt()] = it }
        authenticationData?.let { connectProps[MQTT5PropertyType.AUTHENTICATION_DATA.toInt()] = it }
        properties?.let { connectProps.putAll(it) }
        
        val propsBytes = MQTT5PropertyEncoder.encodeProperties(connectProps)
        val propsLen = MQTTProtocol.encodeRemainingLength(propsBytes.size)
        variableHeader.addAll(propsLen.toList())
        variableHeader.addAll(propsBytes.toList())
        
        val payload = mutableListOf<Byte>()
        payload.addAll(MQTTProtocol.encodeString(clientId).toList())
        payload.add(0x00) // Will Properties length = 0
        username?.let { payload.addAll(MQTTProtocol.encodeString(it).toList()) }
        password?.let { payload.addAll(MQTTProtocol.encodeString(it).toList()) }
        
        val remLen = variableHeader.size + payload.size
        val fixed = mutableListOf<Byte>()
        fixed.add(MQTTMessageType.CONNECT)
        fixed.addAll(MQTTProtocol.encodeRemainingLength(remLen).toList())
        
        return (fixed + variableHeader + payload).toByteArray()
    }
    
    fun buildConnackV5(
        reasonCode: Int = MQTT5ReasonCode.SUCCESS,
        sessionPresent: Boolean = false,
        properties: Map<Int, Any>? = null
    ): ByteArray {
        val variableHeader = mutableListOf<Byte>()
        variableHeader.add(if (sessionPresent) 0x01 else 0x00)
        variableHeader.add(reasonCode.toByte())
        
        val props = properties ?: emptyMap()
        val propsBytes = MQTT5PropertyEncoder.encodeProperties(props)
        val propsLen = MQTTProtocol.encodeRemainingLength(propsBytes.size)
        variableHeader.addAll(propsLen.toList())
        variableHeader.addAll(propsBytes.toList())
        
        val remLen = variableHeader.size
        val fixed = mutableListOf<Byte>()
        fixed.add(MQTTMessageType.CONNACK)
        fixed.addAll(MQTTProtocol.encodeRemainingLength(remLen).toList())
        
        return (fixed + variableHeader).toByteArray()
    }
    
    fun parseConnackV5(data: ByteArray, offset: Int = 0): Triple<Boolean, Int, Map<Int, Any>> {
        if (offset + 2 > data.size) throw IllegalArgumentException("Insufficient data for CONNACK")
        val sessionPresent = (data[offset].toInt() and 0x01) != 0
        val reasonCode = data[offset + 1].toInt() and 0xFF
        var pos = offset + 2
        
        val (propLen, propLenBytes) = MQTTProtocol.decodeRemainingLength(data, pos)
        pos += propLenBytes
        val (props, _) = MQTT5PropertyEncoder.decodeProperties(data.copyOfRange(pos, pos + propLen), 0)
        pos += propLen
        
        return Triple(sessionPresent, reasonCode, props)
    }
    
    fun buildPublishV5(
        topic: String,
        payload: ByteArray,
        packetId: Int? = null,
        qos: Int = 0,
        retain: Boolean = false,
        properties: Map<Int, Any>? = null
    ): ByteArray {
        var msgType = MQTTMessageType.PUBLISH.toInt()
        if (qos > 0) msgType = msgType or (qos shl 1)
        if (retain) msgType = msgType or 0x01
        
        val vh = mutableListOf<Byte>()
        vh.addAll(MQTTProtocol.encodeString(topic).toList())
        if (qos > 0 && packetId != null) {
            vh.add((packetId shr 8).toByte())
            vh.add((packetId and 0xFF).toByte())
        }
        
        val props = properties ?: emptyMap()
        val propsBytes = MQTT5PropertyEncoder.encodeProperties(props)
        val propsLen = MQTTProtocol.encodeRemainingLength(propsBytes.size)
        vh.addAll(propsLen.toList())
        vh.addAll(propsBytes.toList())
        
        val pl = (vh + payload.toList()).toByteArray()
        val remLen = pl.size
        return byteArrayOf(
            msgType.toByte(),
            *MQTTProtocol.encodeRemainingLength(remLen),
            *pl
        )
    }
    
    fun buildSubscribeV5(
        packetId: Int,
        topic: String,
        qos: Int = 0,
        subscriptionIdentifier: Int? = null,
        properties: Map<Int, Any>? = null
    ): ByteArray {
        val vh = mutableListOf<Byte>()
        vh.add((packetId shr 8).toByte())
        vh.add((packetId and 0xFF).toByte())
        
        val props = mutableMapOf<Int, Any>()
        subscriptionIdentifier?.let { props[MQTT5PropertyType.SUBSCRIPTION_IDENTIFIER.toInt()] = it }
        properties?.let { props.putAll(it) }
        val propsBytes = MQTT5PropertyEncoder.encodeProperties(props)
        val propsLen = MQTTProtocol.encodeRemainingLength(propsBytes.size)
        vh.addAll(propsLen.toList())
        vh.addAll(propsBytes.toList())
        
        val pl = MQTTProtocol.encodeString(topic) + byteArrayOf((qos and 0x03).toByte())
        val rem = vh.size + pl.size
        return byteArrayOf(
            (MQTTMessageType.SUBSCRIBE.toInt() or 0x02).toByte(),
            *MQTTProtocol.encodeRemainingLength(rem),
            *vh.toByteArray(),
            *pl
        )
    }
    
    fun buildSubackV5(
        packetId: Int,
        reasonCodes: List<Int>,
        properties: Map<Int, Any>? = null
    ): ByteArray {
        val vh = mutableListOf<Byte>()
        vh.add((packetId shr 8).toByte())
        vh.add((packetId and 0xFF).toByte())
        
        val props = properties ?: emptyMap()
        val propsBytes = MQTT5PropertyEncoder.encodeProperties(props)
        val propsLen = MQTTProtocol.encodeRemainingLength(propsBytes.size)
        vh.addAll(propsLen.toList())
        vh.addAll(propsBytes.toList())
        
        val pl = reasonCodes.map { it.toByte() }.toByteArray()
        val rem = vh.size + pl.size
        return byteArrayOf(
            MQTTMessageType.SUBACK,
            *MQTTProtocol.encodeRemainingLength(rem),
            *vh.toByteArray(),
            *pl
        )
    }
    
    fun parseSubackV5(data: ByteArray, offset: Int = 0): Triple<Int, List<Int>, Map<Int, Any>> {
        if (offset + 2 > data.size) throw IllegalArgumentException("Insufficient data for SUBACK packet ID")
        val pid = ((data[offset].toInt() and 0xFF) shl 8) or (data[offset + 1].toInt() and 0xFF)
        var pos = offset + 2
        
        val (propLen, propLenBytes) = MQTTProtocol.decodeRemainingLength(data, pos)
        pos += propLenBytes
        val (props, _) = MQTT5PropertyEncoder.decodeProperties(data.copyOfRange(pos, pos + propLen), 0)
        pos += propLen
        
        val reasonCodes = mutableListOf<Int>()
        while (pos < data.size) {
            reasonCodes.add(data[pos].toInt() and 0xFF)
            pos++
        }
        
        return Triple(pid, reasonCodes, props)
    }
    
    fun buildUnsubscribeV5(
        packetId: Int,
        topics: List<String>,
        properties: Map<Int, Any>? = null
    ): ByteArray {
        val vh = mutableListOf<Byte>()
        vh.add((packetId shr 8).toByte())
        vh.add((packetId and 0xFF).toByte())
        
        val props = properties ?: emptyMap()
        val propsBytes = MQTT5PropertyEncoder.encodeProperties(props)
        val propsLen = MQTTProtocol.encodeRemainingLength(propsBytes.size)
        vh.addAll(propsLen.toList())
        vh.addAll(propsBytes.toList())
        
        val plList = topics.flatMap { MQTTProtocol.encodeString(it).toList() }
        val pl = ByteArray(plList.size) { plList[it] }
        val rem = vh.size + pl.size
        return byteArrayOf(
            (MQTTMessageType.UNSUBSCRIBE.toInt() or 0x02).toByte(),
            *MQTTProtocol.encodeRemainingLength(rem),
            *vh.toByteArray(),
            *pl
        )
    }
    
    fun buildUnsubackV5(
        packetId: Int,
        reasonCodes: List<Int>? = null,
        properties: Map<Int, Any>? = null
    ): ByteArray {
        val vh = mutableListOf<Byte>()
        vh.add((packetId shr 8).toByte())
        vh.add((packetId and 0xFF).toByte())
        
        val props = properties ?: emptyMap()
        val propsBytes = MQTT5PropertyEncoder.encodeProperties(props)
        val propsLen = MQTTProtocol.encodeRemainingLength(propsBytes.size)
        vh.addAll(propsLen.toList())
        vh.addAll(propsBytes.toList())
        
        val pl = reasonCodes?.map { it.toByte() }?.toByteArray() ?: ByteArray(0)
        val rem = vh.size + pl.size
        return byteArrayOf(
            MQTTMessageType.UNSUBACK,
            *MQTTProtocol.encodeRemainingLength(rem),
            *vh.toByteArray(),
            *pl
        )
    }
    
    fun buildDisconnectV5(
        reasonCode: Int = MQTT5ReasonCode.NORMAL_DISCONNECTION_DISC,
        properties: Map<Int, Any>? = null
    ): ByteArray {
        val vh = mutableListOf<Byte>()
        vh.add(reasonCode.toByte())
        
        val props = properties ?: emptyMap()
        val propsBytes = MQTT5PropertyEncoder.encodeProperties(props)
        val propsLen = MQTTProtocol.encodeRemainingLength(propsBytes.size)
        vh.addAll(propsLen.toList())
        vh.addAll(propsBytes.toList())
        
        val remLen = vh.size
        val fixed = mutableListOf<Byte>()
        fixed.add(MQTTMessageType.DISCONNECT)
        fixed.addAll(MQTTProtocol.encodeRemainingLength(remLen).toList())
        
        return (fixed + vh).toByteArray()
    }
}

File: ios/Sources/MqttQuicPlugin/MQTT/MQTT5Protocol.swift
//
// MQTT5Protocol.swift
// MqttQuicPlugin
//
// MQTT 5.0 Protocol implementation. Matches MQTTD mqttd/protocol_v5.py.
//

import Foundation

public final class MQTT5Protocol {
    
    public static func buildConnectV5(
        clientId: String,
        username: String? = nil,
        password: String? = nil,
        keepalive: UInt16 = 20,
        cleanStart: Bool = true,
        sessionExpiryInterval: UInt32? = nil,
        receiveMaximum: UInt16? = nil,
        maximumPacketSize: UInt32? = nil,
        topicAliasMaximum: UInt16? = nil,
        requestResponseInformation: UInt8? = nil,
        requestProblemInformation: UInt8? = nil,
        authenticationMethod: String? = nil,
        authenticationData: Data? = nil,
        properties: [UInt8: Any]? = nil
    ) throws -> Data {
        var variableHeader = Data()
        variableHeader.append(try MQTTProtocol.encodeString("MQTT"))
        variableHeader.append(MQTTProtocolLevel.v5)
        
        var flags: UInt8 = 0
        if cleanStart { flags |= 0x02 }
        if username != nil { flags |= MQTTConnectFlags.username }
        if password != nil { flags |= MQTTConnectFlags.password }
        variableHeader.append(flags)
        variableHeader.append(UInt8((keepalive >> 8) & 0xFF))
        variableHeader.append(UInt8(keepalive & 0xFF))
        
        var connectProps: [UInt8: Any] = [:]
        if let sei = sessionExpiryInterval {
            connectProps[MQTT5PropertyType.sessionExpiryInterval.rawValue] = sei
        }
        if let rm = receiveMaximum {
            connectProps[MQTT5PropertyType.receiveMaximum.rawValue] = rm
        }
        if let mps = maximumPacketSize {
            connectProps[MQTT5PropertyType.maximumPacketSize.rawValue] = mps
        }
        if let tam = topicAliasMaximum {
            connectProps[MQTT5PropertyType.topicAliasMaximum.rawValue] = tam
        }
        if let rri = requestResponseInformation {
            connectProps[MQTT5PropertyType.requestResponseInformation.rawValue] = rri
        }
        if let rpi = requestProblemInformation {
            connectProps[MQTT5PropertyType.requestProblemInformation.rawValue] = rpi
        }
        if let am = authenticationMethod {
            connectProps[MQTT5PropertyType.authenticationMethod.rawValue] = am
        }
        if let ad = authenticationData {
            connectProps[MQTT5PropertyType.authenticationData.rawValue] = ad
        }
        if let props = properties {
            connectProps.merge(props) { (_, new) in new }
        }
        
        let propsBytes = try MQTT5PropertyEncoder.encodeProperties(connectProps)
        let propsLen = try MQTTProtocol.encodeRemainingLength(propsBytes.count)
        variableHeader.append(contentsOf: propsLen)
        variableHeader.append(propsBytes)
        
        var payload = Data()
        payload.append(try MQTTProtocol.encodeString(clientId))
        payload.append(0x00) // Will Properties length = 0 (no will for now)
        if let u = username { payload.append(try MQTTProtocol.encodeString(u)) }
        if let p = password { payload.append(try MQTTProtocol.encodeString(p)) }
        
        let remLen = variableHeader.count + payload.count
        var fixed = Data()
        fixed.append(MQTTMessageType.CONNECT.rawValue)
        fixed.append(contentsOf: try MQTTProtocol.encodeRemainingLength(remLen))
        
        var out = Data()
        out.append(fixed)
        out.append(variableHeader)
        out.append(payload)
        return out
    }
    
    public static func buildConnackV5(
        reasonCode: MQTT5ReasonCode = .success,
        sessionPresent: Bool = false,
        properties: [UInt8: Any]? = nil
    ) throws -> Data {
        var variableHeader = Data()
        variableHeader.append(sessionPresent ? 0x01 : 0x00)
        variableHeader.append(reasonCode.rawValue)
        
        var props: [UInt8: Any] = [:]
        if let p = properties { props = p }
        let propsBytes = try MQTT5PropertyEncoder.encodeProperties(props)
        let propsLen = try MQTTProtocol.encodeRemainingLength(propsBytes.count)
        variableHeader.append(contentsOf: propsLen)
        variableHeader.append(propsBytes)
        
        let remLen = variableHeader.count
        var out = Data()
        out.append(MQTTMessageType.CONNACK.rawValue)
        out.append(contentsOf: try MQTTProtocol.encodeRemainingLength(remLen))
        out.append(variableHeader)
        return out
    }
    
    public static func parseConnackV5(_ data: Data, offset: Int = 0) throws -> (Bool, MQTT5ReasonCode, [UInt8: Any], Int) {
        if offset + 2 > data.count { throw MQTTProtocolError.insufficientData("CONNACK") }
        let sessionPresent = (data[offset] & 0x01) != 0
        let reasonCode = MQTT5ReasonCode(rawValue: data[offset + 1]) ?? .unspecifiedError
        var pos = offset + 2
        
        let (propLen, propLenBytes) = try MQTTProtocol.decodeRemainingLength(data, offset: pos)
        pos += propLenBytes
        let (props, _) = try MQTT5PropertyEncoder.decodeProperties(data.subdata(in: pos..<(pos + propLen)), offset: 0)
        pos += propLen
        
        return (sessionPresent, reasonCode, props, pos)
    }
    
    public static func buildPublishV5(
        topic: String,
        payload: Data,
        packetId: UInt16? = nil,
        qos: UInt8 = 0,
        retain: Bool = false,
        properties: [UInt8: Any]? = nil
    ) throws -> Data {
        var msgType = MQTTMessageType.PUBLISH.rawValue
        if qos > 0 { msgType |= (qos << 1) }
        if retain { msgType |= 0x01 }
        
        var vh = Data()
        vh.append(try MQTTProtocol.encodeString(topic))
        if qos > 0, let pid = packetId {
            vh.append(UInt8((pid >> 8) & 0xFF))
            vh.append(UInt8(pid & 0xFF))
        }
        
        var props: [UInt8: Any] = [:]
        if let p = properties { props = p }
        let propsBytes = try MQTT5PropertyEncoder.encodeProperties(props)
        let propsLen = try MQTTProtocol.encodeRemainingLength(propsBytes.count)
        vh.append(contentsOf: propsLen)
        vh.append(propsBytes)
        
        let pl = vh + payload
        let remLen = pl.count
        var out = Data()
        out.append(msgType)
        out.append(contentsOf: try MQTTProtocol.encodeRemainingLength(remLen))
        out.append(pl)
        return out
    }
    
    public static func buildSubscribeV5(
        packetId: UInt16,
        topic: String,
        qos: UInt8 = 0,
        subscriptionIdentifier: Int? = nil,
        properties: [UInt8: Any]? = nil
    ) throws -> Data {
        var vh = Data()
        vh.append(UInt8((packetId >> 8) & 0xFF))
        vh.append(UInt8(packetId & 0xFF))
        
        var props: [UInt8: Any] = [:]
        if let si = subscriptionIdentifier {
            props[MQTT5PropertyType.subscriptionIdentifier.rawValue] = si
        }
        if let p = properties { props.merge(p) { (_, new) in new } }
        let propsBytes = try MQTT5PropertyEncoder.encodeProperties(props)
        let propsLen = try MQTTProtocol.encodeRemainingLength(propsBytes.count)
        vh.append(contentsOf: propsLen)
        vh.append(propsBytes)
        
        var pl = Data()
        pl.append(try MQTTProtocol.encodeString(topic))
        pl.append(qos & 0x03)
        
        let rem = vh.count + pl.count
        var out = Data()
        out.append(MQTTMessageType.SUBSCRIBE.rawValue | 0x02)
        out.append(contentsOf: try MQTTProtocol.encodeRemainingLength(rem))
        out.append(vh)
        out.append(pl)
        return out
    }
    
    public static func buildSubackV5(
        packetId: UInt16,
        reasonCodes: [MQTT5ReasonCode],
        properties: [UInt8: Any]? = nil
    ) throws -> Data {
        var vh = Data()
        vh.append(UInt8((packetId >> 8) & 0xFF))
        vh.append(UInt8(packetId & 0xFF))
        
        var props: [UInt8: Any] = [:]
        if let p = properties { props = p }
        let propsBytes = try MQTT5PropertyEncoder.encodeProperties(props)
        let propsLen = try MQTTProtocol.encodeRemainingLength(propsBytes.count)
        vh.append(contentsOf: propsLen)
        vh.append(propsBytes)
        
        var pl = Data()
        for rc in reasonCodes {
            pl.append(rc.rawValue)
        }
        
        let rem = vh.count + pl.count
        var out = Data()
        out.append(MQTTMessageType.SUBACK.rawValue)
        out.append(contentsOf: try MQTTProtocol.encodeRemainingLength(rem))
        out.append(vh)
        out.append(pl)
        return out
    }
    
    public static func parseSubackV5(_ data: Data, offset: Int = 0) throws -> (UInt16, [MQTT5ReasonCode], [UInt8: Any], Int) {
        if offset + 2 > data.count { throw MQTTProtocolError.insufficientData("SUBACK packet ID") }
        let pid = (UInt16(data[offset] & 0xFF) << 8) | UInt16(data[offset + 1] & 0xFF)
        var pos = offset + 2
        
        let (propLen, propLenBytes) = try MQTTProtocol.decodeRemainingLength(data, offset: pos)
        pos += propLenBytes
        let (props, _) = try MQTT5PropertyEncoder.decodeProperties(data.subdata(in: pos..<(pos + propLen)), offset: 0)
        pos += propLen
        
        var reasonCodes: [MQTT5ReasonCode] = []
        while pos < data.count {
            reasonCodes.append(MQTT5ReasonCode(rawValue: data[pos]))
            pos += 1
        }
        
        return (pid, reasonCodes, props, pos)
    }
    
    public static func buildUnsubscribeV5(
        packetId: UInt16,
        topics: [String],
        properties: [UInt8: Any]? = nil
    ) throws -> Data {
        var vh = Data()
        vh.append(UInt8((packetId >> 8) & 0xFF))
        vh.append(UInt8(packetId & 0xFF))
        
        var props: [UInt8: Any] = [:]
        if let p = properties { props = p }
        let propsBytes = try MQTT5PropertyEncoder.encodeProperties(props)
        let propsLen = try MQTTProtocol.encodeRemainingLength(propsBytes.count)
        vh.append(contentsOf: propsLen)
        vh.append(propsBytes)
        
        var pl = Data()
        for t in topics {
            pl.append(try MQTTProtocol.encodeString(t))
        }
        
        let rem = vh.count + pl.count
        var out = Data()
        out.append(MQTTMessageType.UNSUBSCRIBE.rawValue | 0x02)
        out.append(contentsOf: try MQTTProtocol.encodeRemainingLength(rem))
        out.append(vh)
        out.append(pl)
        return out
    }
    
    public static func buildUnsubackV5(
        packetId: UInt16,
        reasonCodes: [MQTT5ReasonCode]? = nil,
        properties: [UInt8: Any]? = nil
    ) throws -> Data {
        var vh = Data()
        vh.append(UInt8((packetId >> 8) & 0xFF))
        vh.append(UInt8(packetId & 0xFF))
        
        var props: [UInt8: Any] = [:]
        if let p = properties { props = p }
        let propsBytes = try MQTT5PropertyEncoder.encodeProperties(props)
        let propsLen = try MQTTProtocol.encodeRemainingLength(propsBytes.count)
        vh.append(contentsOf: propsLen)
        vh.append(propsBytes)
        
        var pl = Data()
        if let rcs = reasonCodes {
            for rc in rcs {
                pl.append(rc.rawValue)
            }
        }
        
        let rem = vh.count + pl.count
        var out = Data()
        out.append(MQTTMessageType.UNSUBACK.rawValue)
        out.append(contentsOf: try MQTTProtocol.encodeRemainingLength(rem))
        out.append(vh)
        out.append(pl)
        return out
    }
    
    public static func buildDisconnectV5(
        reasonCode: MQTT5ReasonCode = .normalDisconnectionDisc,
        properties: [UInt8: Any]? = nil
    ) throws -> Data {
        var vh = Data()
        vh.append(reasonCode.rawValue)
        
        var props: [UInt8: Any] = [:]
        if let p = properties { props = p }
        let propsBytes = try MQTT5PropertyEncoder.encodeProperties(props)
        let propsLen = try MQTTProtocol.encodeRemainingLength(propsBytes.count)
        vh.append(contentsOf: propsLen)
        vh.append(propsBytes)
        
        let remLen = vh.count
        var out = Data()
        out.append(MQTTMessageType.DISCONNECT.rawValue)
        out.append(contentsOf: try MQTTProtocol.encodeRemainingLength(remLen))
        out.append(vh)
        return out
    }
}
